File: ios/Pets/ViewModels/Sources/ViewModels/List/ListViewModel.swift
//
//  File.swift
//  
//
//  Created by Dylan Velez on 11/08/21.
//

import Foundation
import Architecture
import Models
import Domain
import Combine

public final class ListViewModel: ViewModel {
    public typealias Event = ListEvent
    public typealias ListStore = Store<ListState, ListEvent, ListReducer>
    
    public struct ViewState {
        public var list: Pets = []
        public var favorites: Pets = []
        public var isLoading: Bool = false
    }
    
    @Published private(set) public var state: ViewState = ViewState()
    public var store: ListStore
    private var stateCancellable: AnyCancellable?
    
    public init(store: ListStore) {
        self.store = store
        stateCancellable = store.$state.sink(receiveValue: stateChanged)
    }
    
    public func stateChanged(newState: ListState) {
        if !newState.list.isEmpty {
            DispatchQueue.main.async {
                let oldList = Set(self.state.list).union(Set(newState.list))
                self.state.list = Array(oldList).sorted { $0.id < $1.id }
            }
        }
        self.state.isLoading = newState.isLoading
        self.state.favorites = newState.favorites
    }
}

File: android/Pets/app/src/main/java/com/pets/viewModels/list/ListViewModel.kt
package com.pets.viewModels.list

import androidx.lifecycle.MutableLiveData
import com.pets.architecture.stores.Store
import com.pets.architecture.viewModels.ViewModel
import com.pets.domain.list.ListEvent
import com.pets.domain.list.ListReducer
import com.pets.domain.list.ListState
import com.pets.models.Pet

class ListViewModel: ViewModel<ListState, ListEvent, ListViewModel.ViewState> {
    data class ViewState(val list: MutableList<Pet> = emptyList<Pet>().toMutableList(),
    val favorites: MutableList<Pet> = emptyList<Pet>().toMutableList(),
    val isLoading: Boolean = false)

    override var state: MutableLiveData<ViewState> = MutableLiveData(ViewState())
    override val store: Store<ListState, ListEvent> =
        Store(MutableLiveData(ListState()), ListReducer())

    override fun onStateChange(newState: ListState) {
        state.postValue(ViewState(list = newState.list,
        isLoading = newState.isLoading,
        favorites = newState.favorites))
    }
}